### PR TITLE
21. Share/Unshare Feature

### LINKED ISSUE #48: 21. Share/Unshare Feature
## Summary

Adding a component with functionality that allows users to invite their friends to their lists. It should allow the user to input a name that they can identify their friend with. Each list should show an icon with a number that represents how many users it is share with. When clicked, a modal will pop up that then allows the user to unshare lists.

## Acceptance criteria
- [x] Create a SharedWithList component
- [x] Create logic that allows user to unshare a list they have shared
- [x] Create a button that shows how many users a list has been shared with
- [x] Update styling to match app/figma file

### PR DESCRIPTION
**Pull request by eonflower**
## Description

- Updated `modal `component to be reusable
- Created a `SharedWithList` component that allows users to see what users that have shared a list with
    - Each user shared with also has a delete button so that users can unshare
    - There is a multi-select / search tool for users to choose multiple lists and also search their lists
    - It will not show lists the user does not own, they can only share their own lists
 - InviteForm allows user to enter a name to attach to the user they are sharing it with so they can better identify later
 - Added a `useSharedWithData` and `unshareList` to the firebase file, and also updated `shareList` function to check if a list has already been shared with a user, and also update the list to have an array of objects that holds the users name and email
 - Decoupled the `capitalize` function to be more reusable

## Related Issue

closes #48 

## Acceptance Criteria

- [x] Create a SharedWithList component
- [x] Create logic that allows user to unshare a list they have shared
- [x] Create a button that shows how many users a list has been shared with
- [x] Update styling to match app/figma file

## Type of Changes

enhancement

### Before
Screenshot 2024-03-29 at 9 39 25 PM

### After
Screenshot 2024-03-29 at 9 39 47 PM
Screenshot 2024-03-29 at 9 40 00 PM
Screenshot 2024-03-29 at 9 41 17 PM

## Testing Steps / QA Criteria

1. git pull from main
2. git checkout an-share-feature
3. npm install
4. navigate to list page
5. if the list is owned by you, a share button should be available
6. share with a user (can use my email to test eonflower.dev@gmail.com)
7. once a list has been shared, it should pop up the icon with a number corresponding to the number of people its been shared with

Some notes as always:
- Previous shared lists will not have the proper schema since I updated lists to gain an array of users (their name and email), so they will not appear in the number with the icon
- I did light styling for shared with page, it wasn't really my focus, but am very open to changing it
- I personally liked having the share button on the list page instead of on the listItem, but we could decouple and make a component for it so that it can be used in multiple places
- I used the [react-select library ](https://react-select.com/home) for the select tool because this was how I envisioned it. I liked it, but also I didn't really ask, so we can remove it if y'all don't like it

### CODE REVIEW

File: src/api/firebase.js [hunk 1]
```diff
@@ -137,30 +138,116 @@ export async function createList(userId, userEmail, listName) {
...
+		if (!listPath) return;
+
+		const fetchSharedWithData = async () => {
+			try {
+				const listDocRef = doc(db, listPath);
+				const listDocSnap = await getDoc(listDocRef);
+
+				if (listDocSnap.exists()) {
+					const listData = listDocSnap.data();
+					const sharedWithData = listData.sharedWith || []; // Assuming sharedWith is an array
```
Comment (StefieCaff) [inline src/api/firebase.js]: That comment is confusing. Is it an array?

File: src/api/firebase.js [hunk 2]
```diff
@@ -137,30 +138,116 @@ export async function createList(userId, userEmail, listName) {
...
+					setSharedWith(sharedWithData);
+				} else {
```
Comment (StefieCaff) [inline src/api/firebase.js:158]: We can add toasts here once that component is done.

File: src/api/firebase.js [hunk 3]
```diff
@@ -137,30 +138,116 @@ export async function createList(userId, userEmail, listName) {
...
+		email: input.recipientEmail,
+	};
+
+	// Share each selected list with the recipient user
+	const sharePromises = selectedLists.map(async (list) => {
+		const listDocumentRef = doc(db, list.value);
+		const userDocumentRef = doc(db, 'users', recipientEmail);
+
+		// Check if the list is already shared with the recipient user
+		const listDoc = await getDoc(listDocumentRef);
```
Comment (StefieCaff) [inline src/api/firebase.js:201]: In the case where the list is already shared, can there be a message to the user here letting them know it was already shared?

File: src/components/Modal.jsx [hunk 4]
```diff
@@ -4,9 +4,13 @@ const Modal = ({ isOpen, onClose, children }) => {
...
-					<div className="modal-content">
-						<button className="close-btn" onClick={onClose}>
+				<div className="fixed inset-0 z-50  xsm:px-4 sm:px-12 overflow-auto flex justify-center items-center backdrop-blur-lg backdrop-filter">
+					<div className="modal-overlay fixed inset-0 bg-black opacity-50"></div>
+					<div className="modal-content relative p-[20px] bg-white rounded-lg shadow-lg">
+						<button
+							className="close-btn absolute top-4 right-6"
+							onClick={onClose}
+						>
 							Close
```
Comment (StefieCaff) [inline src/components/Modal.jsx]: I personally like the 'X' instead of 'close' button...just my vote :)

**Review by StefieCaff — COMMENTED** [review]
Added a few comments, nothing blocking. 

NOICE!!! Really nice :)
I had a couple thoughts --- love how you made the share option only available to lists that are owned. Solves that problem beautifully. Wish I'da thought of that sooner.
The only thing I found that I think should be changed before merge is that when I added the users name that I wanted it associate with the shared list, it saved it as lowercase. I think using the code that was implemented in issue 10 will work?
I prefer in the design how the button is top left-- maybe just not styled yet.. just putting in my 2 cents.

**Comment by StefieCaff** [comment]
I resolved the conflict in App.jsx because it was from my most recent commit.

**Review by StefieCaff — APPROVED** [review]

File: src/api/firebase.js (hunk 3 above)
Comment (eonflower) [reply src/api/firebase.js:201]: I tried implementing this, but couldn't get it to work. Would love a solution if someone else has the brain power.

**Comment by eonflower** [comment]
I updated the placement of the share button and merged it with the delete button too :)

**Pull request merged**